server/routerlicious/packages/services-telemetry/src/lumberjack.ts

```diff
@@ -0,0 +1,111 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "path";
+import { LumberEventName } from "./lumberEventNames";
+import { Lumber } from "./lumber";
+import { LogLevel, LumberType, ITelemetryMetadata, ILumberjackEngine } from "./resources";
+
+// Lumberjack is a telemetry manager class that allows the collection of metrics and logs
+// throughout the service. A list of ILumberjackEngine must be provided to Lumberjack
+// by calling setupEngines() before Lumberjack can be used.
+export class Lumberjack {
+    protected static _instance: Lumberjack | undefined = undefined;
+    private readonly engineList: ILumberjackEngine[];
+    private isSetupCompleted: boolean;
+
+    protected constructor() {
+        this.engineList = [];
+        this.isSetupCompleted = false;
+    }
+
+    public static get instance() {
+        if (!Lumberjack._instance) {
+            Lumberjack._instance = new Lumberjack();
+        }
+
+        return Lumberjack._instance;
+    }
+
+    public setupEngines(engines: ILumberjackEngine[]) {
+        if (this.isSetupCompleted) {
+            throw new Error("This Lumberjack was already setup with a list of engines.");
```

Comment:
Yes, in this case my idea was to disallow overriding the engines. The idea is similar to what we currently do with Winston: we initially setup the Winston transports we want to use. Winston might not have this restriction of preventing new transports to be added eventually, but we never do it. Ideally, the setup should be done in the beginning, when the process starts running, and the same set of engines would exist throughout the whole process duration - avoiding mistakes of logs/metrics suddenly disappearing or going somewhere else because somebody accidentally overwrote the engine list. I though about it as some sort of constructor parameter - but since  I made Lumberjack a singleton, it is awkward to accept constructor parameters. Hence the setupEngines() method.

If this is an anti-pattern or there are legitimate cases when we want to allow new engines to be added throughout the process, let me know and I can update the method above 😄

Reply:
That makes a lot of sense! I don't think it's an antipattern, and it sounds well thought out. Thanks for explaining!